src/parser/paladin/protection/modules/items/LucidDreams.tsx

```diff
@@ -0,0 +1,125 @@
+import ItemStatistic from 'interface/statistics/ItemStatistic';
+import SpellLink from 'common/SpellLink';
+import { formatPercentage } from 'common/format';
+import SpellUsable from '../features/SpellUsable';
+import Seraphim from '../talents/Seraphim';
+
+class LucidDreams extends Analyzer {
+  static dependencies = {
+    abilities: Abilities,
+    spellUsable: SpellUsable,
+    sera: Seraphim,
+  };
+
+  protected abilities!: Abilities;
+  protected spellUsable!: SpellUsable;
+  protected sera!: Seraphim;
+
+  constructor(options: any) {
+    super(options);
+    this.active = this.selectedCombatant.hasEssence(SPELLS.LUCID_DREAMS.traitId);
+
+    if (!this.active) {
+      return;
+    }
+
+    if (this.selectedCombatant.hasMajor(SPELLS.LUCID_DREAMS.traitId)) {
+      options.abilities.add({
+        spell: SPELLS.LUCID_DREAMS_MAJOR,
+        category: Abilities.SPELL_CATEGORIES.ITEMS,
+        cooldown: 120,
+        gcd: {
+          base: 1500,
+        },
+        castEfficiency: {
+          suggestion: true,
+          recommendedEfficiency: 0.9,
+        },
+      });
+      this.addEventListener(Events.applybuff.to(SELECTED_PLAYER).spell(SPELLS.LUCID_DREAMS_MAJOR), this.onMajorStart);
+      this.addEventListener(Events.removebuff.to(SELECTED_PLAYER).spell(SPELLS.LUCID_DREAMS_MAJOR), this.onMajorEnd);
+    }
+    this.addEventListener(Events.cast.by(SELECTED_PLAYER).spell(SPELLS.SHIELD_OF_THE_RIGHTEOUS), this.onCastSotR);
+    this.addEventListener(Events.cast.by(SELECTED_PLAYER).spell(SPELLS.SERAPHIM_TALENT), this.onCastSotR);
+    // the heal occurs when a refund happens
+    this.addEventListener(Events.heal.to(SELECTED_PLAYER).spell(SPELLS.LUCID_DREAMS_HEAL), this.onRefund);
+  }
+
+  private lastCast?: CastEvent;
+  private refundedCD = {
+    [SPELLS.SERAPHIM_TALENT.id]: 0,
+    [SPELLS.SHIELD_OF_THE_RIGHTEOUS.id]: 0,
+  };
+
+  private refundCounts = {
+    [SPELLS.SERAPHIM_TALENT.id]: 0,
+    [SPELLS.SHIELD_OF_THE_RIGHTEOUS.id]: 0,
+  };
+
+  onMajorStart(event: ApplyBuffEvent) { }
+  onMajorEnd(event: RemoveBuffEvent) { }
+
+  onCastSotR(event: CastEvent) {
+    this.lastCast = event;
+  }
+
+  onRefund(event: HealEvent) {
+    let refund;
+    if (this.lastCast!.ability.guid === SPELLS.SERAPHIM_TALENT.id) {
+      refund = this.sera.lastCDConsumed / 2;
+    } else {
+      refund = this.abilities.getExpectedCooldownDuration(SPELLS.SHIELD_OF_THE_RIGHTEOUS.id, event)! / 2;
+    }
+    try {
+      this.spellUsable.reduceCooldown(SPELLS.SHIELD_OF_THE_RIGHTEOUS.id, refund);
+    } catch (e) {
+      // i have found exactly 1 log where this throws an error because
+      // SotR is not on CD, but I cannot figure out why. The refund
+      // events are always after the cast events (and looking at
+      // fabricated events in the event view confirms that).
+      //
+      // it only happens once in that one log, so I'm going to just move on
+      console.error(e);
+      console.log(this.lastCast);
+    }
```

Comment:
Since this only happened in that one log then maybe it's fine, but console.log in production :(